File: day2/main.go
package main

import (
	"flag"
	"fmt"
	"github.com/Piszmog/adventofcode/utils"
	"log"
	"strconv"
	"strings"
)

func main() {
	passwordsFile := flag.String("f", "passwords.txt", "File containing the passwords")

	var passwords []password
	err := utils.ReadTextFile(*passwordsFile, func(line string, rowNumber int) error {
		lineParts := strings.Split(line, " ")
		// get the min/max
		minMax := strings.Split(lineParts[0], "-")
		min, err := strconv.Atoi(minMax[0])
		if err != nil {
			return fmt.Errorf("failed to convert min %s: %w", minMax[0], err)
		}
		max, err := strconv.Atoi(minMax[1])
		if err != nil {
			return fmt.Errorf("failed to convert max %s: %w", minMax[1], err)
		}
		// get the letter
		letter := lineParts[1][:1]
		// get the value
		value := lineParts[2]
		passwords = append(passwords, password{
			min:    min,
			max:    max,
			letter: letter,
			value:  value,
		})
		return nil
	})
	if err != nil {
		log.Fatalln(err)
	}

	// Part 1
	validPasswords1 := getValidPasswordsPart1(passwords)
	fmt.Printf("Part 1: There are %d valid passwords\n", len(validPasswords1))

	// Part 2
	validPasswords2 := getValidPasswordsPart2(passwords)
	fmt.Printf("Part 2: There are %d valid passwords\n", len(validPasswords2))
}

type password struct {
	min    int
	max    int
	letter string
	value  string
}

func getValidPasswordsPart1(passwords []password) []password {
	var validPasswords []password
	for _, password := range passwords {
		count := strings.Count(password.value, password.letter)
		if count >= password.min && count <= password.max {
			validPasswords = append(validPasswords, password)
		}
	}
	return validPasswords
}

func getValidPasswordsPart2(passwords []password) []password {
	var validPasswords []password
	for _, password := range passwords {
		value := password.value
		r := []rune(password.letter)[0]
		hasPosition1 := false
		hasPosition2 := false
		if len(value) >= password.min {
			hasPosition1 = rune(value[password.min-1]) == r
		}
		if len(value) >= password.max {
			hasPosition2 = rune(value[password.max-1]) == r
		}
		if (hasPosition1 && !hasPosition2) || (!hasPosition1 && hasPosition2) {
			validPasswords = append(validPasswords, password)
		}
	}
	return validPasswords
}

File: day2/rust/src/main.rs
use std::fs::File;
use std::io::{BufRead, BufReader};

fn main() {
    let passwords = get_passwords();

    // part 1
    let valid_password_count = get_valid_passwords_count_1(&passwords);
    println!("Part 1: There are {} valid passwords", valid_password_count);

    // part 2
    let valid_password_count = get_valid_passwords_count_2(&passwords);
    println!("Part 2: There are {} valid passwords", valid_password_count);
}

struct Password {
    min: i32,
    max: i32,
    letter: String,
    value: String,
}

fn get_passwords() -> Vec<Password> {
    let file = File::open("../passwords.txt").expect("could not find file");
    let reader = BufReader::new(file);
    reader.lines()
        .map(|line| line.expect("could not read line"))
        .map(|line| {
            let parts: Vec<&str> = line.split(" ").collect();
            let min_max: Vec<&str> = parts[0].split("-").collect();
            let min = min_max[0].parse::<i32>().unwrap();
            let max = min_max[1].parse::<i32>().unwrap();
            let letter = parts[1].trim_end_matches(":").to_string();
            Password {
                min,
                max,
                letter,
                value: parts[2].to_string(),
            }
        })
        .collect()
}

fn get_valid_passwords_count_1(passwords: &Vec<Password>) -> i32 {
    let mut valid_password_count = 0;
    passwords.iter().for_each(|p| {
        let count = p.value.matches(&p.letter).count();
        if count >= p.min as usize && count <= p.max as usize {
            valid_password_count += 1;
        }
    });
    valid_password_count
}

fn get_valid_passwords_count_2(passwords: &Vec<Password>) -> i32 {
    let mut valid_password_count = 0;
    passwords.iter().for_each(|p| {
        let r = p.letter.chars().next().unwrap();

        let has_pos_1 = p.value.len() >= p.min as usize
            && p.value.chars().nth((p.min - 1) as usize).unwrap() == r;

        let has_pos_2 = p.value.len() >= p.max as usize
            && p.value.chars().nth((p.max - 1) as usize).unwrap() == r;

        if (has_pos_1 && !has_pos_2) || (!has_pos_1 && has_pos_2) {
            valid_password_count += 1;
        }
    });
    valid_password_count
}
